[RelayMiner] LeanClient (#567)
pokt-network/poktroll: docusaurus/docs/operate/configs/relayminer_config.md

## Global options

```yaml
default_signing_key_names: [ <string>, <string> ]
smt_store_path: <string>
```

### `default_signing_key_names`

_`Required`_ if `suppliers.*.signing_key_names` is not specified. 

> This configuration option specifies a list of key names from the keyring that the `RelayMiner` will use to sign transactions. These keys are used to derive the public key and the corresponding address, which are essential for interacting with the Pocket network. Each key name listed here must be present in the keyring used to start the `RelayMiner` instance.

Reviewer: Multiline (80-120) plaease

> This configuration option specifies a list of key names from the keyring that the `RelayMiner` will use to sign transactions. These keys are used to derive the public key and the corresponding address, which are essential for interacting with the Pocket network. Each key name listed here must be present in the keyring used to start the `RelayMiner` instance.

Reviewer: ```suggestion
This configuration option specifies a list of key names from the keyring that the `RelayMiner` will use to sign transactions. These key names are used to derive the public key and the corresponding address, which are essential for interacting with POKT network. Each key name listed here must be present in the keyring used to start the `RelayMiner` instance.
```

#### Example Configuration
```yaml
default_signing_key_names:
  - keyname1
  - keyname2
```

In this example, `keyname1` and `keyname2` are the names of keys in the keyring that will be used to sign transactions.

#### How It Works

> - If a supplier does not specify its own `signing_key_names`, the supplier will use the keys listed in `default_signing_key_names`.

Reviewer: Consider adding a flow chart to make this visual: https://mermaid.js.org/syntax/flowchart.html

I bet you can ask ChatGPT to create one based on the text as a starting point.

Author: I've restructured the documentation page - I think it is more user-friendly now. Also includes the diagram.

- You can specify more than one key, allowing the `RelayMiner` to operate multiple suppliers with different keys/addresses.

> - This setup is useful for running multiple RelayMiner instances across different regions or for redundancy.

Reviewer: This seems like a benefit, not a how it works.

    S2 -.- RM4
```

> TODO(#528): Currently, it is not possible to run a single RelayMiner instance with multiple `signing_key_names` due to the complexity in determining which key to use for each supplier, especially if suppliers have overlapping services. This feature will be available once issue #528 is resolved.

Reviewer: Aren't you resolving this?

It MUST match the `Service.Id` specified by the supplier when staking for the
service.

### `signing_key_names`

_`Required`_ if `default_signing_key_names` is empty.

> This option specifies the list of signing key names specific to a supplier. If a supplier does not provide its own `signing_key_names`, the `RelayMiner` will use the `default_signing_key_names`.

Reviewer: ```suggestion
This option specifies the list of signing key names specific to a supplier.
If a supplier does not provide its own `signing_key_names`, the `RelayMiner` will use the `default_signing_key_names`.
```

#### Example Configuration

```yaml
default_signing_key_names:
  - supplier_key2
  - supplier_key3
suppliers:
  # This supplier will only run for `supplier_key1` and `supplier_key4` keys

>   - service_id: supplier1

Reviewer: I'm a bit confused:

1. Can you avoid using `supplier` for `service_id`?
2. Do you mean `This supplier will ...` or `This relayminder will ...`

In either case, can you try to put yourself in the shoes of someone who has never thought about this and make it really obvious how this works? Even I'm confused...

Maybe add a mermaid diagram if its applicable.

Author: 1. 👍
2. Reworded as "RelayMiner will only provide service for `llm-model` using `supplier_key1` and `supplier_key4` keys."

    signing_key_names:
      - supplier_key1
      - supplier_key4
    # ... the rest of the config

  # This supplier will only run for `supplier_key2` and `supplier_key3` keys
  - service_id: supplier2
    signing_key_names: []
    # ... the rest of the config
```
#### How It Works

- If `signing_key_names` are specified for a supplier, the `RelayMiner` will use these keys instead of the `default_signing_key_names`.
- This allows each supplier to have a unique set of keys for signing transactions, providing flexibility and better key management.

#### Benefits

- **Granular Control**: Allows each supplier to have its own specific keys, enhancing security and management.

> - **Fallback Mechanism**: If signing_key_names are not specified, the RelayMiner falls back to `default_signing_key_names`, ensuring continuity of operations.

Reviewer: ```suggestion
- **Fallback Mechanism**: If `signing_key_names` are not specified, the RelayMiner falls back to `default_signing_key_names`, ensuring continuity of operations.
```

### Example Configuration

```yaml
default_signing_key_names:
  - keyname1
  - keyname2
suppliers:
  # RelayMiner will only provide service for `llm-model` using `keyname3` and `keyname4` keys.
  - service_id: llm-model
    signing_key_names:
      - keyname3
      - keyname4
    # ... the rest of the config

  # RelayMiner will only provide service for `etherium-mainnet` using `keyname1` and `keyname2` keys.
  - service_id: etherium-mainnet
    # ... the rest of the config
```

```mermaid
flowchart TB
    subgraph Global_Config
        direction TB
        A[default_signing_key_names] --> B[keyname1]
        A --> C[keyname2]
    end

    subgraph llm-model-config
        direction TB
        D[service_id: llm-model]
        D --> E[signing_key_names]
        E --> F[keyname3]
        E --> G[keyname4]
    end

    subgraph etherium-mainnet-config
        direction TB
        H[service_id: etherium-mainnet]
        H --> I[signing_key_names]
        I --> J[Not Specified]
    end

    subgraph Effective_Signing_Keys
        direction TB
        K[llm-model] --> L[keyname3]
        K --> M[keyname4]

>         N[Supplier 2] --> O[keyname1]

Reviewer: ```suggestion
        N[etherium-mainnet] --> O[keyname1]
```
